Mathlib/Analysis/Analytic/Constructions.lean

```diff
@@ -137,6 +137,87 @@ theorem AnalyticOn.curry_right {f : E × F → G} {x : E} {s : Set (E × F)} (fa
   fun y m ↦ (fa (x, y) m).along_snd
 alias AnalyticOn.along_snd := AnalyticOn.curry_right
 
+/-!
+### Analyticity in Pi spaces
+-/
+
+section
+
+variable {ι : Type*} [Fintype ι] {e : E} {Fm : ι → Type*}
+    [∀ i, NormedAddCommGroup (Fm i)] [∀ i, NormedSpace 𝕜 (Fm i)]
+    {f : Π i, E → Fm i} {s : Set E} {r : ℝ≥0∞}
+    {p : Π i, FormalMultilinearSeries 𝕜 E (Fm i)}
+
+lemma FormalMultilinearSeries.le_pi_radius (h : ∀ i, r ≤ (p i).radius) :
+    r ≤ (FormalMultilinearSeries.pi p).radius := by
+  apply le_of_forall_nnreal_lt (fun r' hr' ↦ ?_)
+  have I i : ∃ C > 0, ∀ n, ‖p i n‖ * (r' : ℝ) ^ n ≤ C :=
+    norm_mul_pow_le_of_lt_radius _ (hr'.trans_le (h i))
+  choose C C_pos hC using I
+  obtain ⟨D, D_nonneg, hD⟩ : ∃ D ≥ 0, ∀ i, C i ≤ D :=
+    ⟨∑ i, C i, Finset.sum_nonneg (fun i _ ↦ (C_pos i).le),
+      fun i ↦ Finset.single_le_sum (fun j _ ↦ (C_pos j).le) (Finset.mem_univ _)⟩
+  apply le_radius_of_bound _ D (fun n ↦ ?_)
+  rcases le_or_lt ((r' : ℝ)^n) 0 with hr' | hr'
+  · exact le_trans (mul_nonpos_of_nonneg_of_nonpos (by positivity) hr') D_nonneg
+  · simp only [pi]
+    rw [← le_div_iff₀ hr', ContinuousMultilinearMap.opNorm_pi,
+      pi_norm_le_iff_of_nonneg (by positivity)]
+    intro i
+    exact (le_div_iff₀ hr').2 ((hC i n).trans (hD i))
+
+lemma HasFPowerSeriesWithinOnBall.pi
```

Comment:
Here and below: is it hard to prove the `Iff` version? Am I right that `0 < r` is automatic for nonempty `ι`?

Reply:
I have proved a few of the iff versions. However, for the last statements (analyticWithinAt and friends), #16752 is needed if I want a very short proof. So you can either merge this one now (and I'll add the missing bits to an API-extending PR which is scheduled after #16752) or merge #16752 now and then I will add the missing bits to the current PR.

Reply:
Merged #16752